File: src/com/company/java_examples/DepthFirstSearch.java
package com.company.java_examples;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class DepthFirstSearch {

    // Leetcode 1971

    public static void main(String[] args) {
        // todo test on leetcode
    }

    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer, Set<Integer>> map = getMap(edges);
        return depthFirstSearch(map, source, destination, new HashSet<>());
    }

    private static Map<Integer, Set<Integer>> getMap(int[][] edges) {
        Map<Integer, Set<Integer>> map = new HashMap<>();
        for (int[] edge : edges) {
            int source = edge[0];
            int destination = edge[1];
            Set<Integer> sourceAdjacent = map.getOrDefault(source, new HashSet<>());
            Set<Integer> destinationAdjacent = map.getOrDefault(destination, new HashSet<>());
            sourceAdjacent.add(destination);
            destinationAdjacent.add(source);
            map.put(source, sourceAdjacent);
            map.put(destination, destinationAdjacent);
        }
        return map;
    }

    private static boolean depthFirstSearch(Map<Integer, Set<Integer>> map, int current, int destination, Set<Integer> visited) {
        if (current == destination) {
            return true;
        }
        if (visited.contains(current)) {
            return false;
        }
        visited.add(current);
        for (Integer child : map.get(current)) {
            if (depthFirstSearch(map, child, destination, visited)) {
                return true;
            }
        }
        return false;
    }
}

File: src/com/company/kotlin_examples/DepthFirstSearch.kt
package com.company.kotlin_examples

object DepthFirstSearch {

    // Leetcode 1971
    @JvmStatic
    fun main(args: Array<String>) {
        // todo test on leetcode
    }

    fun validPath(n: Int, edges: Array<Array<Int>>, source: Int, destination: Int): Boolean {
        val map = getMap(edges)
        return depthFirstSearch(map, source, destination, java.util.HashSet())
    }

    private fun getMap(edges: Array<Array<Int>>): Map<Int, MutableSet<Int>> {
        val map: MutableMap<Int, MutableSet<Int>> = HashMap()
        for (edge in edges) {
            val source = edge[0]
            val destination = edge[1]
            val sourceAdjacent = map.getOrDefault(source, HashSet())
            val destinationAdjacent = map.getOrDefault(destination, HashSet())
            sourceAdjacent.add(destination)
            destinationAdjacent.add(source)
            map[source] = sourceAdjacent
            map[destination] = destinationAdjacent
        }
        return map
    }

    private fun depthFirstSearch(
        map: Map<Int, MutableSet<Int>>,
        current: Int,
        destination: Int,
        visited: MutableSet<Int>
    ): Boolean {
        if (current == destination) {
            return true
        }
        if (visited.contains(current)) {
            return false
        }
        visited.add(current)
        for (child in map[current]!!) {
            if (depthFirstSearch(map, child, destination, visited)) {
                return true
            }
        }
        return false
    }
}